### PR TITLE
PROTOTECH-14: use mulDiv to allow higher debt in pool when calculating LUP

### PR DESCRIPTION
**Pull request by grandizzy**
# Description of change
## High level
* See https://github.com/ajna-finance/contracts/pull/856
  * `LUP` computing overflows when there's big debt in pool, in Prototech regression test `72911661459413347380086.629833841230757427`. While this isn't something regular, it could happen for strange tokens. Changed the way scaled sum is calculated from `(runningScale * scaling * value + 5e35) / 1e36` to use `Math.mulDiv` in order to keep precision and to allow higher values of debt

# Description of bug or vulnerability and solution
* Allow higher debt in pools
* Use `Math.wmul` in order to keep precision and to allow computing LUP for higher debt amounts 

# Contract size
## Pre Change
```
============ Deployment Bytecode Sizes ============
  ERC721Pool         -  24,476B  (99.59%)
  ERC20Pool          -  23,710B  (96.47%)
```
## Post Change
```
============ Deployment Bytecode Sizes ============
  ERC721Pool               -  24,404B  (99.30%)
  ERC20Pool                -  23,638B  (96.18%)
```

# Gas usage
## Pre Change
```
| Function Name                        | min             | avg    | median | max     | # calls |
| addQuoteToken                        | 122519          | 179833 | 164382 | 715582  | 60004   |
| bucketTake                           | 310035          | 328676 | 328676 | 347317  | 4       |
| drawDebt                             | 253325          | 298341 | 272221 | 812976  | 136003  |
| kick                                 | 149519          | 223592 | 222084 | 1035153 | 48000   |
| kickWithDeposit                      | 200288          | 277971 | 274117 | 1001101 | 8000    |
| moveQuoteToken                       | 250002          | 250002 | 250002 | 250002  | 1       |
| removeQuoteToken                     | 148134          | 172220 | 176922 | 196623  | 4000    |
| repayDebt                            | 577752          | 609908 | 599800 | 758170  | 16002   |
| settle                               | 351196          | 351196 | 351196 | 351196  | 1       |
| take                                 | 82491           | 83841  | 83513  | 316305  | 15998   |
```
## Post Change
```
| Function Name                        | min             | avg    | median | max     | # calls |
| addQuoteToken                        | 122519          | 179809 | 164382 | 715426  | 60004   |
| bucketTake                           | 309660          | 328301 | 328302 | 346943  | 4       |
| drawDebt                             | 253325          | 298329 | 272221 | 812768  | 136003  |
| kick                                 | 149467          | 223488 | 221980 | 1035049 | 48000   |
| kickWithDeposit                      | 199768          | 277399 | 273545 | 1000997 | 8000    |
| moveQuoteToken                       | 270918          | 270918 | 270918 | 270918  | 1       |
| removeQuoteToken                     | 147562          | 171692 | 176350 | 198515  | 4000    |
| repayDebt                            | 577596          | 609752 | 599644 | 758014  | 16002   |
| settle                               | 350739          | 350739 | 350739 | 350739  | 1       |
| take                                 | 82116           | 83693  | 83429  | 315931  | 15998   |
```

### CODE REVIEW

File: src/libraries/internal/Deposits.sol [hunk 1]
```diff
@@ -99,7 +101,13 @@ library Deposits {
             // Compute sum up to sumIndex_ + i
             scaledValue =
                 lowerIndexSum +
-                (scaling != 0 ?  (runningScale * scaling * value + 5e35) / 1e36 : Maths.wmul(runningScale, value));
+                (
+                    scaling != 0 ? Math.mulDiv(
```
Comment (grandizzy) [inline src/libraries/internal/Deposits.sol:107]: this rounds down by default, not sure if should round up or if any big of difference. Tests shows same scaling discrepancy of 10 in both cases

File: tests/forge/unit/ERC721Pool/ERC721PoolEMAs.t.sol [hunk 2]
```diff
@@ -190,7 +190,7 @@ contract ERC721PoolEMAsTest is ERC721HelperContract {
             rate:   0.0405 * 1e18                   // dropping at 4.05%
         });
         (, , , depositEma) = _pool.emasInfo();
-        assertEq(depositEma, 17_664.344669688571758689 * 1e18);         // still moving toward 20_000
+        assertEq(depositEma, 17_664.344669688571758688 * 1e18);         // still moving toward 20_000
```
Comment (grandizzy) [inline tests/forge/unit/ERC721Pool/ERC721PoolEMAs.t.sol:193]: this test doesn't change if using Rounding Up

File: tests/forge/unit/ERC721Pool/ERC721PoolEMAs.t.sol [hunk 3]
```diff
@@ -199,7 +199,7 @@ contract ERC721PoolEMAsTest is ERC721HelperContract {
             rate:   0.03645 * 1e18                  // second interest rate drop
         });                  
         (, , , depositEma) = _pool.emasInfo();
-        assertEq(depositEma, 19_855.532581473734007629 * 1e18);         // reached (sort of) 20_000
+        assertEq(depositEma, 19_855.532581473734007628 * 1e18);         // reached (sort of) 20_000
```
Comment (grandizzy) [inline tests/forge/unit/ERC721Pool/ERC721PoolEMAs.t.sol:202]: this test doesn't change if using Rounding Up

File: tests/forge/unit/FenwickTree.t.sol [hunk 4]
```diff
@@ -130,9 +130,9 @@ contract FenwickTreeTest is DSTestPlus {
         uint256 subMax = Maths.max(Maths.wmul(subIndexSum, factor), _tree.prefixSum(subIndex));
         uint256 subMin = Maths.min(Maths.wmul(subIndexSum, factor), _tree.prefixSum(subIndex));
 
-        // 3 >= scaling discrepency
-        assertLe(max - min, 3);
-        assertLe(subMax - subMin, 3);
+        // 10 >= scaling discrepency
```
Comment (grandizzy) [inline tests/forge/unit/FenwickTree.t.sol:133]: scaling discrepancy increase to same 10 in both Rounding Down/Rounding Up cases

File: src/libraries/internal/Deposits.sol (hunk 1 above)
Comment (prat-gpt) [reply src/libraries/internal/Deposits.sol:107]: I think rounding down is in favour of the pool, but not sure how this will affect the index calculations.

Comment (mattcushman) [reply src/libraries/internal/Deposits.sol:107]: Here, rounding depends on how it's being used I think, but we never pushed down context dependent rounding to this point.  The old rounds nearest, new rounds down.

**Review by mattcushman — APPROVED** [review]
lgtm.  nice to see that it has such a small effect on baselines

File: tests/forge/regression/ERC20Pool/prototech/PrototechRegression.t.sol [hunk 5]
```diff
@@ -274,6 +274,107 @@ contract PrototechRegressionTestLiquidationWith50BucketsIndex1ERC20Pool is Liqui
         super.setUp();
     }
 
+    function test_regression_prototech_R14() external {
```
Comment (ith-harvey) [inline tests/forge/regression/ERC20Pool/prototech/PrototechRegression.t.sol:282]: Add comments please : )

**Review by prat-gpt — APPROVED** [review]
LGTM!

**Review by ith-harvey — APPROVED** [review]

**Pull request merged**